pkg/utils/github_utils.go

```diff
@@ -2,29 +2,94 @@ package utils
 
 import (
 	"context"
+	"encoding/base64"
+	"fmt"
+	"net/url"
+	"os"
 	"time"
 
 	"github.com/google/go-github/v59/github"
+	"golang.org/x/oauth2"
 )
 
-// GetLatestGitHubRepoRelease returns the latest release tag for a GitHub repository
-func GetLatestGitHubRepoRelease(owner string, repo string) (string, error) {
-	opt := &github.ListOptions{Page: 1, PerPage: 1}
-	client := github.NewClient(nil)
+// newGitHubClient creates a new GitHub client. If a token is provided, it returns an authenticated client;
+// otherwise, it returns an unauthenticated client.
+func newGitHubClient(ctx context.Context) *github.Client {
+	githubToken := os.Getenv("GITHUB_TOKEN")
+	if githubToken == "" {
+		return github.NewClient(nil)
+	}
+
+	// Token found, create an authenticated client
+	ts := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: githubToken},
+	)
+	tc := oauth2.NewClient(ctx, ts)
 
-	ctx, cancel := context.WithTimeout(context.Background(), time.Second*2)
+	return github.NewClient(tc)
+}
+
+// GetLatestGitHubRepoRelease returns the latest release tag for a GitHub repository.
+func GetLatestGitHubRepoRelease(owner string, repo string) (string, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
 	defer cancel()
 
+	client := newGitHubClient(ctx)
+	opt := &github.ListOptions{Page: 1, PerPage: 1}
+
 	releases, _, err := client.Repositories.ListReleases(ctx, owner, repo, opt)
 	if err != nil {
-		return "", err
+		return "", fmt.Errorf("failed to list releases: %w", err)
 	}
 
-	if len(releases) > 0 {
-		latestRelease := releases[0]
-		latestReleaseTag := *latestRelease.TagName
-		return latestReleaseTag, nil
+	if len(releases) > 0 && releases[0].TagName != nil {
+		return *releases[0].TagName, nil
 	}
 
 	return "", nil
 }
+
+// DownloadFileFromGitHub downloads a file from a GitHub repository using the GitHub API.
+func DownloadFileFromGitHub(rawURL string) ([]byte, error) {
```

Comment:
This might be good, but it's changing the implementation in a risky way. The original request was simply to add a bearer token header, not change the implementation to use the GitHub API.

Reply:
@osterman func DownloadFileFromGitHub(rawURL string) ([]byte, error) - this is a whole new function that i added, its being used in the new functionality I proposed for vednoring, atmos.yaml with env var and in stacks with folder flag. Its not used anywhere else. The existion function func GetLatestGitHubRepoRelease used github API before me and I just added a func to configure a client with auth token to it (nitive functionality). So the one above is a whole new, if we dont need that right now for vendoring, stacks and atmos.yaml, i can take the fucn out.  Let me know whats best.